INSURANCE-PROVIDER
INSURANCE-PROVIDER-ADDRESS
INSURANCE-POLICY-NUMBER
INSURANCE-GROUP-NUMBER
INSURANCE-GROUP-NAME


In [11]:
import random
import re

In [1]:
insurance_templates = [
    "Insurer - {INSURANCE-PROVIDER} | Address - {INSURANCE-PROVIDER-ADDRESS} | Policy - {INSURANCE-POLICY-NUMBER} | Group - {INSURANCE-GROUP-NUMBER} | Group Name - {INSURANCE-GROUP-NAME}",
    "Provider: {INSURANCE-PROVIDER} - Location: {INSURANCE-PROVIDER-ADDRESS} - Policy: {INSURANCE-POLICY-NUMBER} - Group ID: {INSURANCE-GROUP-NUMBER} - Group: {INSURANCE-GROUP-NAME}",
    "Insurance: {INSURANCE-PROVIDER}, Address: {INSURANCE-PROVIDER-ADDRESS}, Policy Number: {INSURANCE-POLICY-NUMBER}, Group: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Policy: {INSURANCE-POLICY-NUMBER} | Insurer: {INSURANCE-PROVIDER} | Address: {INSURANCE-PROVIDER-ADDRESS} | Group: {INSURANCE-GROUP-NUMBER} | Group Name: {INSURANCE-GROUP-NAME}",
    "Insurance Company: {INSURANCE-PROVIDER}, Office Location: {INSURANCE-PROVIDER-ADDRESS}, Policy ID: {INSURANCE-POLICY-NUMBER}, Group ID: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Insurer - {INSURANCE-PROVIDER}, Location - {INSURANCE-PROVIDER-ADDRESS}, Policy No. - {INSURANCE-POLICY-NUMBER}, Group ID - {INSURANCE-GROUP-NUMBER}, Group - {INSURANCE-GROUP-NAME}",
    "Company: {INSURANCE-PROVIDER}, Addr: {INSURANCE-PROVIDER-ADDRESS}, Policy: {INSURANCE-POLICY-NUMBER}, Group ID: {INSURANCE-GROUP-NUMBER}, Group: {INSURANCE-GROUP-NAME}",
    "Insurer: {INSURANCE-PROVIDER} | Addr: {INSURANCE-PROVIDER-ADDRESS} | Policy No.: {INSURANCE-POLICY-NUMBER} | Group ID: {INSURANCE-GROUP-NUMBER} | Group: {INSURANCE-GROUP-NAME}",
    "Policy Number: {INSURANCE-POLICY-NUMBER}, Insurer: {INSURANCE-PROVIDER}, Address: {INSURANCE-PROVIDER-ADDRESS}, Group ID: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Insurance Org: {INSURANCE-PROVIDER}, Office: {INSURANCE-PROVIDER-ADDRESS}, Policy ID: {INSURANCE-POLICY-NUMBER}, Group ID: {INSURANCE-GROUP-NUMBER}, Group: {INSURANCE-GROUP-NAME}",
    "Provider: {INSURANCE-PROVIDER} - Location: {INSURANCE-PROVIDER-ADDRESS} - Policy No.: {INSURANCE-POLICY-NUMBER} - Group ID: {INSURANCE-GROUP-NUMBER} - Group: {INSURANCE-GROUP-NAME}",
    "Insurer: {INSURANCE-PROVIDER} - Address: {INSURANCE-PROVIDER-ADDRESS} - Policy No.: {INSURANCE-POLICY-NUMBER} - Group ID: {INSURANCE-GROUP-NUMBER} - Group: {INSURANCE-GROUP-NAME}",
    "Policy Number: {INSURANCE-POLICY-NUMBER}, Insurer: {INSURANCE-PROVIDER}, Address: {INSURANCE-PROVIDER-ADDRESS}, Group ID: {INSURANCE-GROUP-NUMBER}, Group: {INSURANCE-GROUP-NAME}",
    "Company Name: {INSURANCE-PROVIDER}, Location: {INSURANCE-PROVIDER-ADDRESS}, Policy: {INSURANCE-POLICY-NUMBER}, Group: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Insurance Provider: {INSURANCE-PROVIDER} - Provider Address: {INSURANCE-PROVIDER-ADDRESS} - Policy Number: {INSURANCE-POLICY-NUMBER} - Group Number: {INSURANCE-GROUP-NUMBER} - Group Name: {INSURANCE-GROUP-NAME}",
    "Insurer: {INSURANCE-PROVIDER}, Address: {INSURANCE-PROVIDER-ADDRESS}, Policy No.: {INSURANCE-POLICY-NUMBER}, Group: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Insurance Provider: {INSURANCE-PROVIDER}, Provider Location: {INSURANCE-PROVIDER-ADDRESS}, Policy ID: {INSURANCE-POLICY-NUMBER}, Group ID: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Provider Name: {INSURANCE-PROVIDER}, Provider Addr: {INSURANCE-PROVIDER-ADDRESS}, Policy No.: {INSURANCE-POLICY-NUMBER}, Group ID: {INSURANCE-GROUP-NUMBER}, Group Name: {INSURANCE-GROUP-NAME}",
    "Insurer: {INSURANCE-PROVIDER} | Location: {INSURANCE-PROVIDER-ADDRESS} | Policy Number: {INSURANCE-POLICY-NUMBER} | Group ID: {INSURANCE-GROUP-NUMBER} | Group Name: {INSURANCE-GROUP-NAME}",
    "Company: {INSURANCE-PROVIDER} - Address: {INSURANCE-PROVIDER-ADDRESS} - Policy: {INSURANCE-POLICY-NUMBER} - Group: {INSURANCE-GROUP-NUMBER} - Group Name: {INSURANCE-GROUP-NAME}",
    "Provider: {INSURANCE-COMPANY} - Address: {INSURANCE-PROVIDER-ADDRESS} - Policy #: {INSURANCE-POLICY-NUMBER}",
    "Group: {INSURANCE-GROUP-NAME} - Policy Number: {INSURANCE-POLICY-NUMBER} - Insurer: {INSURANCE-PROVIDER}",
    "Work Location: {INSURANCE-PROVIDER-ADDRESS} - Policy Number (Group): {INSURANCE-POLICY-NUMBER} ({INSURANCE-GROUP-NUMBER})",
    "Company: {INSURANCE-PROVIDER} - Policy#: {INSURANCE-POLICY-NUMBER} - Group Name: {INSURANCE-GROUP-NAME}",
    "Policy Number: {INSURANCE-POLICY-NUMBER} - Group Name: {INSURANCE-GROUP-NAME} - Insurance Organization: {INSURANCE-PROVIDER}",
    "Insurance Co.: {INSURANCE-PROVIDER} - Address: {INSURANCE-PROVIDER-ADDRESS} - Group#: {INSURANCE-GROUP-NUMBER}",
    "Policy#: {INSURANCE-POLICY-NUMBER} - Location: {INSURANCE-PROVIDER-ADDRESS} - Group: {INSURANCE-GROUP-NAME}",
    "Insurance Provider: {INSURANCE-PROVIDER} (Claim#: 123456) - Policy Number: {INSURANCE-POLICY-NUMBER} - Group: {INSURANCE-GROUP-NAME}",
    "Address: {INSURANCE-PROVIDER-ADDRESS} - Policy Number: {INSURANCE-POLICY-NUMBER} - Group (Name): {INSURANCE-GROUP-NAME}",
    "Insurer: {INSURANCE-PROVIDER} - Group Name: {INSURANCE-GROUP-NAME} - Policy#: {INSURANCE-POLICY-NUMBER}",
    "Policy Number: {INSURANCE-POLICY-NUMBER} - Insurance Company: {INSURANCE-PROVIDER} - Group (Number): {INSURANCE-GROUP-NUMBER}",
    "Location: {INSURANCE-PROVIDER-ADDRESS} - Group: {INSURANCE-GROUP-NAME} - Policy#: {INSURANCE-POLICY-NUMBER}",
    "Insurance Provider: {INSURANCE-PROVIDER} (Claim#: 789012) - Policy Number: {INSURANCE-POLICY-NUMBER} - Group Name: {INSURANCE-GROUP-NAME}",
    "Work Address: {INSURANCE-PROVIDER-ADDRESS} - Policy Number (Group): {INSURANCE-POLICY-NUMBER} ({INSURANCE-GROUP-NUMBER})",
    "Group: {INSURANCE-GROUP-NAME} - Policy#: {INSURANCE-POLICY-NUMBER} - Insurance Co.: {INSURANCE-PROVIDER} (Written Note: Contact John Smith)",
    "Claim#: 123456 - Insurance Provider: {INSURANCE-PROVIDER} - Policy#: {INSURANCE-POLICY-NUMBER} - Group: {INSURANCE-GROUP-NAME}",
    "Insurance Organization: {INSURANCE-PROVIDER} (Claim#: 987654) - Policy Number: {INSURANCE-POLICY-NUMBER} - Group Name: {INSURANCE-GROUP-NAME}",
    "Address: {INSURANCE-PROVIDER-ADDRESS} - Group Name: {INSURANCE-GROUP-NAME} - Policy Number: {INSURANCE-POLICY-NUMBER}",
    "Insurance Co.: {INSURANCE-PROVIDER} - Location: {INSURANCE-PROVIDER-ADDRESS} - Policy#: {INSURANCE-POLICY-NUMBER}",
    "Insurer Name: {INSURANCE-PROVIDER}, Address: {INSURANCE-PROVIDER-ADDRESS}, Policy ID: {INSURANCE-POLICY-NUMBER}",
    "Insurance Provider: {INSURANCE-PROVIDER}, Location: {INSURANCE-PROVIDER-ADDRESS}, Policy Number: {INSURANCE-POLICY-NUMBER}, Group ID: {INSURANCE-GROUP-NUMBER}",
    "Insurance Organization: {INSURANCE-PROVIDER} || Address: {INSURANCE-PROVIDER-ADDRESS} || Policy #: {INSURANCE-POLICY-NUMBER}",
    "{INSURANCE-PROVIDER} Insurance Company, {INSURANCE-PROVIDER-ADDRESS} Address, Policy No {INSURANCE-POLICY-NUMBER}, Group {INSURANCE-GROUP-NUMBER}",
    "Insurance Co: {INSURANCE-PROVIDER}, Work Location: {INSURANCE-PROVIDER-ADDRESS}, Policy: {INSURANCE-POLICY-NUMBER",
    "{INSURANCE-PROVIDER} | {INSURANCE-PROVIDER-ADDRESS} | Policy Number {INSURANCE-POLICY-NUMBER} | Group Name: {INSURANCE-GROUP-NAME}",
    "Insurance Provider: {INSURANCE-PROVIDER} Address: {INSURANCE-PROVIDER-ADDRESS} Policy Number: {INSURANCE-POLICY-NUMBER}",
    "{INSURANCE-PROVIDER}, {INSURANCE-PROVIDER-ADDRESS}, Policy: {INSURANCE-POLICY-NUMBER}, Group: {INSURANCE-GROUP-NUMBER}, Name: {INSURANCE-GROUP-NAME}",
    "Insurance Company Name: {INSURANCE-PROVIDER} Location: {INSURANCE-PROVIDER-ADDRESS} Policy ID: {INSURANCE-POLICY-NUMBER} Additional Text Additional Text",
    "{INSURANCE-PROVIDER} - {INSURANCE-PROVIDER-ADDRESS} - Policy {INSURANCE-POLICY-NUMBER} - Group {INSURANCE-GROUP-NUMBER} - {INSURANCE-GROUP-NAME}",
    "Insurance Organization: {INSURANCE-PROVIDER} Policy Number: {INSURANCE-POLICY-NUMBER} Group: {INSURANCE-GROUP-NUMBER} Name: {INSURANCE-GROUP-NAME",
    "{INSURANCE-PROVIDER} {INSURANCE-PROVIDER-ADDRESS} Policy: {INSURANCE-POLICY-NUMBER}",
    "Insurance Company: {INSURANCE-PROVIDER} Address: {INSURANCE-PROVIDER-ADDRESS} Policy #: {INSURANCE-POLICY-NUMBER} Group: {INSURANCE-GROUP-NUMBER} Additional Text",
    "Insurance Provider Name: {INSURANCE-PROVIDER} Address: {INSURANCE-PROVIDER-ADDRESS} Policy ID: {INSURANCE-POLICY-NUMBER} Group ID: {INSURANCE-GROUP-NUMBER} Group Name: {INSURANCE-GROUP-NAME}",
    "{INSURANCE-PROVIDER} - Policy {INSURANCE-POLICY-NUMBER} - Group {INSURANCE-GROUP-NUMBER} - {INSURANCE-GROUP-NAME}",
    "Insurance Company: {INSURANCE-PROVIDER} Location: {INSURANCE-PROVIDER-ADDRESS} Policy Number: {INSURANCE-POLICY-NUMBER",
    "{INSURANCE-PROVIDER} | {INSURANCE-PROVIDER-ADDRESS} | Policy: {INSURANCE-POLICY-NUMBER} | Group: {INSURANCE-GROUP-NUMBER}",
    "Insurance Provider: {INSURANCE-PROVIDER} Address: {INSURANCE-PROVIDER-ADDRESS} Extra Text Policy Number: {INSURANCE-POLICY-NUMBER}",
    "{INSURANCE-PROVIDER} || {INSURANCE-PROVIDER-ADDRESS} || Policy: {INSURANCE-POLICY-NUMBER} || Group Name: {INSURANCE-GROUP-NAME}",
    "Insurance Company Name: {INSURANCE-PROVIDER} Address: {INSURANCE-PROVIDER-ADDRESS} Policy Number: {INSURANCE-POLICY-NUMBER} Group: {INSURANCE-GROUP-NUMBER} Additional Text Here",
]

In [2]:
from faker import Faker

fake = Faker()


def generate_insurance_provider():
    return fake.company()


def generate_insurance_provider_address():
    return fake.address().replace("\n", " ")


def generate_insurance_policy_number():
    return fake.random_element(
        elements=(
            fake.bothify(text="???######"),
            fake.bothify(text="P-???#####"),
            fake.bothify(text="ID########"),
        )
    )


def generate_insurance_group_number():
    return fake.random_element(
        elements=(
            fake.bothify(text="GRP-???######"),
            fake.bothify(text="G-???#####"),
            fake.bothify(text="GID########"),
        )
    )


def generate_insurance_group_name():
    return fake.company() + " Group"

In [5]:
placeholder_to_generator = {
    "INSURANCE-PROVIDER": generate_insurance_provider,
    "INSURANCE-PROVIDER-ADDRESS": generate_insurance_provider_address,
    "INSURANCE-POLICY-NUMBER": generate_insurance_policy_number,
    "INSURANCE-GROUP-NUMBER": generate_insurance_group_number,
    "INSURANCE-GROUP-NAME": generate_insurance_group_name,
}

In [15]:
def replace_placeholders(template, placeholder_to_generator):
    updated_text = template
    entities = []

    # Iterate over each placeholder in the template text
    while "{" in updated_text:
        # Find the next placeholder in the updated text
        match = re.search(r"\{([^{}]*)\}", updated_text)
        if not match:
            break
        placeholder = match.group(1)

        # Replace the placeholder with the generated value
        fake_value = str(placeholder_to_generator[placeholder]())
        start, end = match.span()
        updated_text = updated_text[:start] + fake_value + updated_text[end:]

        # # Update the start and end positions for entities
        # print("Placeholder:", placeholder)
        # print("Start:", start)
        # print("End:", start + len(fake_value))
        entities.append((start, start + len(fake_value), placeholder))

    return updated_text, {"entities": entities}

In [16]:
def generate_training_data(placeholder_to_generator, templates, num_samples):
    training_data = []

    for _ in range(num_samples):
        template = random.choice(templates)
        updated_text, entities = replace_placeholders(
            template, placeholder_to_generator
        )
        training_data.append((updated_text, entities))

    return training_data

In [20]:
# Example usage
from typing import Tuple

num_samples = 1
training_data = []
training_data = generate_training_data(
    placeholder_to_generator, insurance_templates, num_samples
)
print(training_data[0])
sample = training_data[0]

('Insurer - Macdonald-Clark | Address - 7607 Black Isle Suite 365 North Justinton, MI 94090 | Policy - 818978 | Group - 1496 | Group Name - Carlson-Hebert Group', {'entities': [(10, 25, 'INSURANCE-PROVIDER'), (38, 89, 'INSURANCE-PROVIDER-ADDRESS'), (101, 107, 'INSURANCE-POLICY-NUMBER'), (118, 122, 'INSURANCE-GROUP-NUMBER'), (138, 158, 'INSURANCE-GROUP-NAME')]})


Insurer - Macdonald-Clark | Address - 7607 Black Isle Suite 365 North Justinton, MI 94090
